File: examples/pattern-core-python/standard_graph.py
"""StandardGraph Python example using the public relateby.pattern API."""

from relateby.pattern import Pattern, StandardGraph, StringVal, Subject


def main():
    parsed = StandardGraph.from_gram("(eve:Person)-[:KNOWS]->(frank:Person)")
    print(f"Parsed graph nodes: {parsed.node_count}")

    # --- Graph construction from native Pattern[Subject] values ---
    alice = Pattern.point(
        Subject.from_id("alice").with_label("Person").with_property("name", StringVal("Alice"))
    )
    bob = Pattern.point(
        Subject.from_id("bob").with_label("Person").with_property("name", StringVal("Bob"))
    )
    relationship = Pattern(
        value=Subject.from_id("r1").with_label("KNOWS"),
        elements=[alice, bob],
    )

    g = StandardGraph.from_patterns([relationship])
    print(f"Nodes: {g.node_count}")           # 2
    print(f"Relationships: {g.relationship_count}")  # 1
    print(f"Source of r1: {g.source('r1').value.identity}")  # alice
    print(f"Target of r1: {g.target('r1').value.identity}")  # bob
    print(f"Neighbors of alice: {[node.value.identity for node in g.neighbors('alice')]}")

    # --- Iteration ---
    print("\nAll nodes:")
    for node_id, pattern in g.nodes():
        print(f"  {node_id}: {sorted(pattern.value.labels)}")

    print("\nAll relationships:")
    for rel_id, rel in g.relationships():
        print(f"  {rel_id}: {rel['source']} -> {rel['target']}")

    print("\n✓ StandardGraph Python example complete")


if __name__ == "__main__":
    main()

File: examples/archive/pattern-core-wasm/standard_graph.mjs
/**
 * StandardGraph WASM/Node.js example (T042)
 *
 * Demonstrates using the StandardGraph binding from Node.js.
 * Run after ci-local.sh builds the wasm-node package:
 *   node examples/pattern-core-wasm/standard_graph.mjs
 *
 * The wasm-node build uses CommonJS; imports via createRequire.
 */

import { createRequire } from "module";
const require = createRequire(import.meta.url);
const wasm = require("../../typescript/@relateby/pattern/wasm-node/pattern_wasm.js");
const { StandardGraph, WasmSubject: Subject } = wasm;

// --- Build a small social graph ---
const g = new StandardGraph();

const alice = new Subject("alice", ["Person"], { name: "Alice" });
const bob = new Subject("bob", ["Person"], { name: "Bob" });
const rel = new Subject("r1", ["KNOWS"], {});

g.addNode(alice);
g.addNode(bob);
// Pass Subject objects directly — no need to spell out identity strings
g.addRelationship(rel, alice, bob);

console.log(`Node count: ${g.nodeCount}`);                        // 2
console.log(`Relationship count: ${g.relationshipCount}`);        // 1
console.log(`Source of r1: ${g.source("r1")}`);
console.log(`Target of r1: ${g.target("r1")}`);

// --- fromGram ---
const g2 = StandardGraph.fromGram("(carol:Person {name:'Carol'})-[:KNOWS]->(dave:Person {name:'Dave'})");
console.log(`fromGram node count: ${g2.nodeCount}`);
console.log(`fromGram relationship count: ${g2.relationshipCount}`);

// --- Subject.build() fluent API ---
const builder = Subject.build("eve");
builder.label("Person");
builder.property("name", "Eve");
const eve = builder.done();
g.addNode(eve);
console.log(`After adding Eve: ${g.nodeCount} nodes`);

console.log("✓ StandardGraph WASM example complete");
